plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/snapshot/LinstorVMSnapshotStrategy.java

```diff
@@ -0,0 +1,347 @@
+import java.util.List;
+
+import com.cloud.agent.api.VMSnapshotTO;
+import com.cloud.event.EventTypes;
+import com.cloud.event.UsageEventUtils;
+import com.cloud.storage.DiskOfferingVO;
+import com.cloud.storage.VolumeVO;
+import com.cloud.storage.dao.DiskOfferingDao;
+import com.cloud.storage.dao.VolumeDao;
+import com.cloud.uservm.UserVm;
+import com.cloud.utils.exception.CloudRuntimeException;
+import com.cloud.utils.fsm.NoTransitionException;
+import com.cloud.vm.UserVmVO;
+import com.cloud.vm.VirtualMachine;
+import com.cloud.vm.dao.UserVmDao;
+import com.cloud.vm.snapshot.VMSnapshot;
+import com.cloud.vm.snapshot.VMSnapshotVO;
+import com.cloud.vm.snapshot.dao.VMSnapshotDao;
+import org.apache.cloudstack.engine.subsystem.api.storage.StrategyPriority;
+import org.apache.cloudstack.storage.datastore.db.PrimaryDataStoreDao;
+import org.apache.cloudstack.storage.datastore.db.StoragePoolVO;
+import org.apache.cloudstack.storage.datastore.util.LinstorUtil;
+import org.apache.cloudstack.storage.to.VolumeObjectTO;
+import org.apache.cloudstack.storage.vmsnapshot.DefaultVMSnapshotStrategy;
+import org.apache.cloudstack.storage.vmsnapshot.VMSnapshotHelper;
+import org.apache.log4j.Logger;
+import org.springframework.stereotype.Component;
+
+@Component
+public class LinstorVMSnapshotStrategy extends DefaultVMSnapshotStrategy {
+    private static final Logger log = Logger.getLogger(LinstorVMSnapshotStrategy.class);
+
+    @Inject
+    private VMSnapshotHelper _vmSnapshotHelper;
+    @Inject
+    private UserVmDao _userVmDao;
+    @Inject
+    private VMSnapshotDao vmSnapshotDao;
+    @Inject
+    private VolumeDao volumeDao;
+    @Inject
+    private DiskOfferingDao diskOfferingDao;
+    @Inject
+    private PrimaryDataStoreDao _storagePoolDao;
+
+    private void linstorCreateMultiSnapshot(
+        DevelopersApi api, VMSnapshotVO vmSnapshotVO, List<VolumeObjectTO> volumeTOs)
+        throws ApiException {
+        CreateMultiSnapshotRequest cmsReq = new CreateMultiSnapshotRequest();
+        for (VolumeObjectTO vol : volumeTOs) {
+            Snapshot snap = new Snapshot();
+            snap.setName(vmSnapshotVO.getName());
+            snap.setResourceName(LinstorUtil.RSC_PREFIX + vol.getPath());
+            log.debug(String.format("Add volume %s;%s to snapshot", vol.getName(), snap.getResourceName()));
+            cmsReq.addSnapshotsItem(snap);
+        }
+        log.debug(String.format("Creating multi snapshot %s", vmSnapshotVO.getName()));
+        ApiCallRcList answers = api.createMultiSnapshot(cmsReq);
+        log.debug(String.format("Created multi snapshot %s", vmSnapshotVO.getName()));
+        if (answers.hasError()) {
+            throw new CloudRuntimeException(
+                "Error creating vm snapshots: " + LinstorUtil.getBestErrorMessage(answers));
+        }
+    }
+
+    @Override
+    public VMSnapshot takeVMSnapshot(VMSnapshot vmSnapshot) {
+        log.info("LinstorVMSnapshotStrategy take snapshot");
```

Comment:
```suggestion
        log.info("Take snapshot");
```

class name will be logged